src/tokens/ERC20Vault.sol

```diff
@@ -0,0 +1,161 @@
+                                IMMUTABLES
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice The underlying token the Vault accepts.
+    ERC20 public immutable underlying;
+
+    /// @notice The base unit of the underlying token and hence vault.
+    /// @dev Equal to 10 ** decimals. Used for fixed point arithmetic.
+    uint256 public immutable baseUnit;
+
+    /// @notice Creates a new Vault that accepts a specific underlying token.
+    /// @param _underlying The ERC20 compliant token the Vault should accept.
+    /// @param _namePrefix A name prefix to be applied before
+    /// @param _symbolPrefix The ERC20 compliant token the Vault should accept.
+
+    constructor(
+        ERC20 _underlying,
+        string memory _namePrefix,
+        string memory _symbolPrefix
+    )
+        ERC20(
+            string(abi.encodePacked(_namePrefix, _underlying.name(), " Vault")),
+            string(abi.encodePacked(_symbolPrefix, _underlying.symbol())),
+            _underlying.decimals()
+        )
+    {
+        underlying = _underlying;
+
+        baseUnit = 10**decimals;
+
+        // Prevent minting of shares until
+        // the initialize function is called.
+        totalSupply = type(uint256).max;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        DEPOSIT/WITHDRAWAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Emitted after a successful deposit.
+    /// @param user The address that deposited into the Vault.
+    /// @param underlyingAmount The amount of underlying tokens that were deposited.
+    event Deposit(address indexed user, uint256 underlyingAmount);
+
+    /// @notice Emitted after a successful withdrawal.
+    /// @param owner The address that withdrew from the Vault.
+    /// @param to The destination for withdrawn tokens.
+    /// @param underlyingAmount The amount of underlying tokens that were withdrawn.
+    event Withdraw(address indexed owner, address indexed to, uint256 underlyingAmount);
+
+    /// @notice Deposit a specific amount of underlying tokens.
+    /// @param to The address to receive shares corresponding to the deposit
+    /// @param underlyingAmount The amount of the underlying token to deposit.
+    function deposit(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        // We don't allow depositing 0 to prevent emitting a useless event.
+        require(underlyingAmount != 0, "AMOUNT_CANNOT_BE_ZERO");
+
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+        // Determine the equivalent amount of shares and mint them.
+        _mint(to, shares);
+
+        emit Deposit(to, underlyingAmount);
+
+        // Transfer in underlying tokens from the user.
+        // This will revert if the user does not have the amount specified.
+        underlying.safeTransferFrom(msg.sender, address(this), underlyingAmount);
```

Comment:
instead of doing a delegatecall out to the safeTransfer library, this code can be inlined to save gas

Reply:
it doesnt do a delegatecall, internal libraries get inlined

Reply:
> To realize this in the EVM, code of internal library functions and all functions called from therein will at compile time be included in the calling contract, and a regular JUMP call will be used instead of a DELEGATECALL.

https://docs.soliditylang.org/en/latest/contracts.html#libraries

Reply:
If this was inlined, you would call it by doing SafeTransferLib.safeTransferFrom. Even then, I don't think it would work, because library functions that change state have to be delegate called:

```Library functions can only be called directly (i.e. without the use of DELEGATECALL) if they do not modify the state (i.e. if they are view or pure functions)```

Reply:
thats just not true lol— look at it in a debugger

<img width="1116" alt="Screen Shot 2021-12-29 at 2 49 58 PM" src="https://user-images.githubusercontent.com/26209401/147708960-5945b5ba-1522-4220-a917-750dae270e0a.png">

<img width="1112" alt="Screen Shot 2021-12-29 at 2 50 38 PM" src="https://user-images.githubusercontent.com/26209401/147708972-13756ac8-d15f-4aa5-a376-e2e4cda4f6fd.png">

Reply:
another way you would know it was being delegate-called is because you would have to link it. 

i explicitly disable linking in `.dapprc` so its not even possible here